homeassistant/helpers/entity_component.py

```diff
@@ -50,40 +53,71 @@ def setup(self, config):
         Loads the platforms from the config and will listen for supported
         discovered platforms.
         """
+        run_coroutine_threadsafe(
+            self.async_setup(config), self.hass.loop
+        ).result()
+
+    @asyncio.coroutine
+    def async_setup(self, config):
+        """Set up a full entity component.
+
+        Loads the platforms from the config and will listen for supported
+        discovered platforms.
+
+        This method must be run in the event loop.
+        """
         self.config = config
 
         # Look in config for Domain, Domain 2, Domain 3 etc and load them
         for p_type, p_config in config_per_platform(config, self.domain):
-            self._setup_platform(p_type, p_config)
+            self.hass.async_run_job(self._setup_platform, p_type, p_config)
```

Comment:
You should await till all `_setup_platform` are done (and thus it should be a coroutine). That way we can ensure that at the end of this method all platforms are setup.

You can use `asyncio.gather` after enqueueing them all and then await them. [Example](https://github.com/home-assistant/home-assistant/blob/b8beae9c6c8977cb9bc680eacfb8812488911e8b/homeassistant/components/automation/__init__.py#L347)